File: back-end/scraper.py
import re
import csv
import requests
from bs4 import BeautifulSoup
from urllib.parse import urljoin
from concurrent.futures import ThreadPoolExecutor, as_completed

def scrape_page(url):
    try:
        response = requests.get(url, timeout=5)
        if response.status_code == 200:
            return BeautifulSoup(response.text, 'html.parser')
    except requests.RequestException:
        return None

def find_internal_links(soup, base_url):
    links = []
    for a in soup.find_all('a', href=True):
        link = urljoin(base_url, a['href'])
        if link.startswith(base_url) and is_relevant_link(link) and link not in links:
            links.append(link)
    return links

def is_relevant_link(link):
    irrelevant_extensions = ('.pdf', '.mov', '.jpg', '.png', '.gif', '.zip', '.mp3', '.mp4', '.avi', '.wmv')
    return not link.endswith(irrelevant_extensions)

def extract_emails(soup):
    email_pattern = r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b'
    paragraph_texts = ' '.join([p.get_text() for p in soup.find_all('p')])
    emails = re.findall(email_pattern, paragraph_texts)
    return emails

def main():
    base_url = 'https://www.silvertouch.com/'
    visited_links = set()
    links_to_visit = [base_url]
    found_emails = set()
    max_depth = 3
    max_pages = 100
    current_depth = 0
    pages_visited = 0
    email_goal = 25

    with open('emails.csv', 'w', newline='') as csvfile:
        writer = csv.writer(csvfile)
        writer.writerow(['Email'])

        while links_to_visit and len(found_emails) < email_goal and current_depth <= max_depth:
            next_level_links = []
            with ThreadPoolExecutor(max_workers=10) as executor:
                futures = {executor.submit(scrape_page, link): link for link in links_to_visit if link not in visited_links}

                for future in as_completed(futures):
                    link = futures[future]
                    soup = future.result()
                    if soup:
                        print(f"Scraping {link}....")
                        visited_links.add(link)
                        emails = extract_emails(soup)
                        for email in emails:
                            if email not in found_emails:
                                writer.writerow([email])
                                found_emails.add(email)
                        next_level_links.extend(find_internal_links(soup, base_url))
                        pages_visited += 1
                        if len(found_emails) >= email_goal:
                            break

            links_to_visit = next_level_links
            current_depth += 1

if __name__ == "__main__":
    main()

File: backend/scraper.js
//------------------------------------------- Possible Fixes --------------------------------------------

//To load the library
/*
require(['foo'], function (foo) {
    //foo is now loaded.
});
*/
/*
var axios, cheerio, iconv, fs, csvWriter;
require(['axios'], function(axios){
});
require(['cheerio'], function(cheerio){
});
require(['iconv'], function(iconv){
});
require(['fs'], function(fs){
});
require(['csvWriter'], function(csvWriter){
});
*/
/*

Require Issue: https://requirejs.org/docs/errors.html#notloaded

*/ 
//-------------------------------------------------------------------------------------------------------
const axios = require('axios');
const cheerio = require('cheerio');
const iconv = require('iconv-lite');
const { URL } = require('url');
const fs = require('fs');
const csvWriter = require('csv-write-stream');

class Queue {
    constructor() {
        this.items = [];
    }

    enqueue(element) {
        this.items.push(element);
    }

    dequeue() {
        if (this.isEmpty()) {
            return "Underflow";
        }
        return this.items.shift();
    }

    isEmpty() {
        return this.items.length == 0;
    }

    length() {
        return this.items.length;
    }
}

const visitedLinks = new Set();
const linksToVisit = new Queue();
const foundEmails = new Set();

const writer = csvWriter({ headers: ["Email"] });
writer.pipe(createWriteStream('emails.csv'));

const irrelevantExtensions = ['.pdf', '.mov', '.jpg', '.png', '.gif', '.zip', '.mp3', '.mp4', '.avi', '.wmv', '.m4v', '.jpeg', '.JPG'];
const relevantKeywords = ['contact', 'about', 'team', 'staff', 'support', 'help'];

const CONCURRENCY_LIMIT = 5;
let activeRequests = 0;
let visitedCount = 0;

async function scrapePage(url) {
    try {
        const response = await get(url, { responseType: 'arraybuffer' });
        const decodedContent = decode(response.data, 'utf-8');
        return load(decodedContent);
    } catch (error) {
        console.error(`Failed to scrape ${url}: ${error.message}`);
        return null;
    }
}

function findInternalLinks($, baseUrl) {
    const links = [];
    $('a[href]').each((i, element) => {
        const href = $(element).attr('href');
        if (!href) return;
        const link = new URL(href, baseUrl).toString();
        if (link.startsWith(baseUrl) && isRelevantLink(link)) {
            links.push(link);
        }
    });
    return links;
}

function isRelevantLink(link) {
    if (irrelevantExtensions.some(ext => link.endsWith(ext))) {
        return false;
    }
    if (relevantKeywords.some(keyword => link.toLowerCase().includes(keyword))) {
        return true;
    }
    return !visitedLinks.has(link);
}

function extractEmails($, url) {
    const emailPattern = /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/g;
    const pageText = $('body').text();
    const emails = pageText.match(emailPattern) || [];
    emails.forEach(email => {
        if (!foundEmails.has(email)) {
            foundEmails.add(email);
            writer.write([email]);
        }
    });

    // Obfuscated and hidden emails
    $('script').each((i, element) => {
        const scriptContent = $(element).html();
        // Look for common patterns of obfuscation here and decode them
    });
    
    // Look for mailto: links
    $('a[href^="mailto:"]').each((i, element) => {
        const mailto = $(element).attr('href');
        const email = mailto.replace(/^mailto:/, '');
        if (!foundEmails.has(email)) {
            foundEmails.add(email);
            writer.write([email]);
        }
    });
}

async function processLink(link) {
    if (visitedLinks.has(link)) return;

    console.log(`Scraping ${link}... Visited Count: ${visitedCount}`);
    const $ = await scrapePage(link);
    if (!$) return;

    visitedLinks.add(link);
    visitedCount++; // Increment visited count

    findInternalLinks($, link).forEach(l => linksToVisit.enqueue(l));
    extractEmails($, link);
}

async function main(url) {
    linksToVisit.enqueue(url); // Initialize with the provided URL

    while (linksToVisit.length() > 0 || activeRequests > 0) {
        while (activeRequests < CONCURRENCY_LIMIT && linksToVisit.length() > 0 && visitedCount < 300) {
            const link = linksToVisit.dequeue();
            activeRequests++;
            await processLink(link);
            activeRequests--;
        }
        await new Promise(resolve => setTimeout(resolve, 100)); // Short delay to allow for concurrency
    }
    writer.end();
}

module.exports = main();
